src/Stratis.Bitcoin.Features.Consensus/CoinViews/DBreezeCoinView.cs

```diff
@@ -319,49 +322,92 @@ public Task<RewindData> GetRewindData(int height)
-
-                        foreach (uint256 txId in rewindData.TransactionsToRemove)
+                        for (int height = firstBlockToRewind; height <= currentHeight; height++)
                         {
-                            this.logger.LogTrace("Outputs of transaction ID '{0}' will be removed.", txId);
-                            transaction.RemoveKey("Coins", txId.ToBytes(false));
+                            Row<int, byte[]> row = transaction.Select<int, byte[]>("Rewind", height);
+                            if (!row.Exists)
+                                continue;
+
+                            keysToRemove.Add(row.Key);
+                            var rewindData = this.dBreezeSerializer.Deserialize<RewindData>(row.Value);
+
+                            if (height == (targetHeight + 1))
+                                firstRewindData = rewindData;
+
+                            foreach (uint256 txId in rewindData.TransactionsToRemove)
+                            {
+                                this.logger.LogTrace("Outputs of transaction ID '{0}' will be removed.", txId);
+                                transactionsToRemove.Add(txId);
+                            }
+
+                            foreach (UnspentOutputs coin in rewindData.OutputsToRestore)
+                            {
+                                this.logger.LogTrace("Outputs of transaction ID '{0}' will be restored.", coin.TransactionId);
+                                outputsToRestore.Add(coin);
+                            }
                         }
 
-                        foreach (UnspentOutputs coin in rewindData.OutputsToRestore)
+                        if (keysToRemove.Count > 0)
                         {
-                            this.logger.LogTrace("Outputs of transaction ID '{0}' will be restored.", coin.TransactionId);
-                            transaction.Insert("Coins", coin.TransactionId.ToBytes(false), this.dBreezeSerializer.Serialize(coin.ToCoins()));
+                            transaction.SynchronizeTables("BlockHash", "Coins", "Rewind");
+
+                            var byteListComparer = new ByteListComparer();
+
+                            foreach (int key in keysToRemove.OrderBy(k => k))
+                                transaction.RemoveKey("Rewind", key);
+
+                            foreach (uint256 txId in transactionsToRemove.OrderBy(t => t.ToBytes(false), byteListComparer))
+                                transaction.RemoveKey("Coins", txId.ToBytes(false));
+
+                            foreach (UnspentOutputs coin in outputsToRestore.OrderBy(c => c.TransactionId.ToBytes(false), byteListComparer))
+                                transaction.Insert("Coins", coin.TransactionId.ToBytes(false), this.dBreezeSerializer.Serialize(coin.ToCoins()));
+
+                            this.SetBlockHash(transaction, firstRewindData.PreviousBlockHash);
```

Comment:
This could be useful, I was in a state where I had to rewind lot of height and it took a lot
So I applied this changes and I found a bug: if you have to rewind more than MaxBlocksToRewind, this line fail with because firstRewindData is null

Reply:
Refactored.